### PR TITLE
OHRM5X-1942: Create Claim Expense Type API

### PR DESCRIPTION
**Pull request by ManulMax**

### CODE REVIEW

File: installer/Migration/V5_4_0/Migration.php [hunk 1]
```diff
@@ -63,7 +63,49 @@ public function up(): void
...
+                ->create();
+            $foreignKeyConstraint = new ForeignKeyConstraint(
+                ['added_by'],
+                'ohrm_user',
+                ['id'],
+                'addedByUser',
+                ['onDelete' => 'CASCADE']
+            );
+            $this->getSchemaHelper()->addForeignKey('ohrm_expense_type', $foreignKeyConstraint);
+        }
```
Comment (ManulMax) [inline installer/Migration/V5_4_0/Migration.php:87]: Have to change and add Alter table

File: src/plugins/orangehrmClaimPlugin/Api/ClaimExpenseTypesAPI.php [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+     * @inheritDoc
+     */
+    public function getValidationRuleForGetAll(): ParamRuleCollection
+    {
+        throw $this->getNotImplementedException();
+    }
+
+    /**
+     * @inheritDoc
+     */
```
Comment (ManulMax) [inline src/plugins/orangehrmClaimPlugin/Api/ClaimExpenseTypesAPI.php:70]: Need to add API doc

Comment (RajithaKumara) [reply src/plugins/orangehrmClaimPlugin/Api/ClaimExpenseTypesAPI.php:70]: please add API doc

File: src/plugins/orangehrmClaimPlugin/Api/ClaimExpenseTypesAPI.php [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+    public function getValidationRuleForCreate(): ParamRuleCollection
+    {
+        return new ParamRuleCollection(
+            $this->getValidationDecorator()->requiredParamRule(
+                $this->getNameRule(),
+            ),
```
Comment (RajithaKumara) [inline src/plugins/orangehrmClaimPlugin/Api/ClaimExpenseTypesAPI.php:91]: *

File: src/plugins/orangehrmClaimPlugin/Api/ClaimExpenseTypesAPI.php [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+                    self::PARAMETER_DESCRIPTION,
+                    new Rule(Rules::STRING_TYPE),
+                    new Rule(Rules::LENGTH, [null, self::DESCRIPTION_MAX_LENGTH])
+                ),
+                true
+            ),
+            $this->getValidationDecorator()->notRequiredParamRule(
+                new ParamRule(
+                    self::PARAMETER_STATUS,
+                    new Rule(Rules::BOOL_TYPE)
```
Comment (RajithaKumara) [inline src/plugins/orangehrmClaimPlugin/Api/ClaimExpenseTypesAPI.php:103]:
```suggestion
                    new Rule(Rules::BOOL_VAL)
```
:thinking:

File: src/plugins/orangehrmClaimPlugin/Api/ClaimExpenseTypesAPI.php [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+        );
+        $expenseType->setDescription(
+            $this->getRequestParams()->getStringOrNull(RequestParams::PARAM_TYPE_BODY, self::PARAMETER_DESCRIPTION)
+        );
+        $expenseType->setStatus(
+            $this->getRequestParams()->getBoolean(RequestParams::PARAM_TYPE_BODY, self::PARAMETER_STATUS, true)
+        );
+
+        $userId = $this->getAuthUser()->getUserId();
+        $expenseType->getDecorator()->setUserByUserId($userId);
```
Comment (RajithaKumara) [inline src/plugins/orangehrmClaimPlugin/Api/ClaimExpenseTypesAPI.php:142]:
```suggestion
        $expenseType->getDecorator()->setUserByUserId($this->getAuthUser()->getUserId());
```

File: src/plugins/orangehrmClaimPlugin/Api/ClaimRequestAPI.php [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+     * @return PayGradeService
+     */
+    public function getPayGradeService(): PayGradeService
+    {
+        return $this->getContainer()->get(Services::PAY_GRADE_SERVICE);
+    }
+
+    /**
+     * @inheritDoc
+     * @throws InvalidParamException
```
Comment (RajithaKumara) [inline src/plugins/orangehrmClaimPlugin/Api/ClaimRequestAPI.php:67]:
```suggestion
```

API doc

File: src/plugins/orangehrmClaimPlugin/Api/ClaimRequestAPI.php [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+    public function getValidationRuleForCreate(): ParamRuleCollection
+    {
+        return new ParamRuleCollection(
+            new ParamRule(
+                self::PARAMETER_CLAIM_EVENT_ID,
+                new Rule(Rules::POSITIVE)
+            ),
+            new ParamRule(
+                self::PARAMETER_CURRENCY,
+                new Rule(Rules::REQUIRED)
```
Comment (RajithaKumara) [inline src/plugins/orangehrmClaimPlugin/Api/ClaimRequestAPI.php:88]: are we expecting id?

File: src/plugins/orangehrmClaimPlugin/Api/ClaimRequestAPI.php [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+            }
+
+            if ($this->getPayGradeService()->getPayGradeDao()->getCurrencyById($currencyId) === null) {
+                throw $this->getInvalidParamException(self::PARAMETER_CURRENCY);
+            }
+
+            $claimRequest->setClaimEvent(
+                $this->getClaimService()->getClaimEventDao()->getClaimEventById($claimEventId)
+            );
+            $claimRequest->setCurrency($currencyId);
```
Comment (RajithaKumara) [inline src/plugins/orangehrmClaimPlugin/Api/ClaimRequestAPI.php:132]: This seems wrong. you should set the currency object, not the id. check entity

File: src/plugins/orangehrmClaimPlugin/Api/ClaimRequestAPI.php [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+            $claimRequest->setDescription(
+                $this->getRequestParams()->getStringOrNull(RequestParams::PARAM_TYPE_BODY, self::PARAMETER_REMARKS)
+            );
+
+            $nextId = $this->getClaimService()->getClaimEventDao()->getNextId();
+            $referenceId = date('Ymd') . str_pad("$nextId", 7, 0, STR_PAD_LEFT);
```
Comment (RajithaKumara) [inline src/plugins/orangehrmClaimPlugin/Api/ClaimRequestAPI.php:138]: move this to service method

File: src/plugins/orangehrmClaimPlugin/Api/Model/ClaimExpenseTypesModel.php [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+ * Boston, MA  02110-1301, USA
+ */
+
+namespace OrangeHRM\Claim\Api\Model;
+
+use OrangeHRM\Core\Api\V2\Serializer\ModelTrait;
+use OrangeHRM\Core\Api\V2\Serializer\Normalizable;
+use OrangeHRM\Entity\ExpenseType;
+
+class ClaimExpenseTypesModel implements Normalizable
```
Comment (RajithaKumara) [inline src/plugins/orangehrmClaimPlugin/Api/Model/ClaimExpenseTypesModel.php:26]:
```suggestion
class ClaimExpenseTypeModel implements Normalizable
```

File: src/plugins/orangehrmClaimPlugin/Api/Model/ClaimRequestModel.php [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+                'status',
+            ]
+        );
+        $this->setAttributeNames(
+            [
+                'id',
+                'referenceId',
+                ['claimEvent', 'id'],
+                ['claimEvent', 'name'],
+                'currency',
```
Comment (RajithaKumara) [inline src/plugins/orangehrmClaimPlugin/Api/Model/ClaimRequestModel.php:50]: is this currency id?
better to go like this
```json
currency: {
    id: 
    ...
}
```

File: src/plugins/orangehrmClaimPlugin/Dao/ClaimDao.php [hunk 12]
```diff
@@ -97,4 +99,37 @@ public function getClaimEventCount(ClaimEventSearchFilterParams $claimEventSearc
...
+     * @return int
+     */
+    public function getNextId(): int
+    {
+        $q = $this->createQueryBuilder(ClaimRequest::class, 'request')
+            ->select('MAX(request.id)');
+        $id = $q->getQuery()->execute();
+        $id[0][1]++;
+
+        return $id[0][1];
```
Comment (RajithaKumara) [inline src/plugins/orangehrmClaimPlugin/Dao/ClaimDao.php:133]: want to recheck this

File: src/plugins/orangehrmClaimPlugin/config/routes.yaml [hunk 13]
```diff
@@ -11,3 +11,17 @@ apiv2_claim_event:
   methods: [ GET, PUT]
   defaults:
     _api: OrangeHRM\Claim\Api\ClaimEventAPI
+
+apiv2_claim_expense_types:
+  path: /api/v2/claim/expense/types
```
Comment (RajithaKumara) [inline src/plugins/orangehrmClaimPlugin/config/routes.yaml:16]:
```suggestion
  path: /api/v2/claim/expenses/types
```

File: src/plugins/orangehrmClaimPlugin/config/routes.yaml [hunk 14]
```diff
@@ -11,3 +11,17 @@ apiv2_claim_event:
...
+  controller: OrangeHRM\Core\Controller\Rest\V2\GenericRestController::handle
+  methods: [ POST, GET, DELETE ]
+  defaults:
+    _api: OrangeHRM\Claim\Api\ClaimExpenseTypesAPI
```
Comment (RajithaKumara) [inline src/plugins/orangehrmClaimPlugin/config/routes.yaml:20]:
```suggestion
    _api: OrangeHRM\Claim\Api\ClaimExpenseTypeAPI
```

File: src/plugins/orangehrmClaimPlugin/config/routes.yaml [hunk 15]
```diff
@@ -11,3 +11,17 @@ apiv2_claim_event:
...
+
+apiv2_claim_request:
+  path: /api/v2/claim/request
```
Comment (RajithaKumara) [inline src/plugins/orangehrmClaimPlugin/config/routes.yaml:23]:
```suggestion
  path: /api/v2/claim/requests
```

File: src/plugins/orangehrmClaimPlugin/config/routes.yaml [hunk 16]
```diff
@@ -11,3 +11,17 @@ apiv2_claim_event:
...
     _api: OrangeHRM\Claim\Api\ClaimEventAPI
+
+apiv2_claim_expense_types:
+  path: /api/v2/claim/expense/types
+  controller: OrangeHRM\Core\Controller\Rest\V2\GenericRestController::handle
+  methods: [ POST, GET, DELETE ]
+  defaults:
+    _api: OrangeHRM\Claim\Api\ClaimExpenseTypesAPI
+
+apiv2_claim_request:
```
Comment (RajithaKumara) [inline src/plugins/orangehrmClaimPlugin/config/routes.yaml:22]:
```suggestion
apiv2_claim_requests:
```

File: src/plugins/orangehrmClaimPlugin/entity/ClaimRequest.php [hunk 17]
```diff
@@ -0,0 +1,301 @@
...
+ * You should have received a copy of the GNU General Public License along with this program;
+ * if not, write to the Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA  02110-1301, USA
+ */
+
+namespace OrangeHRM\Entity;
+
+use DateTime;
+use Doctrine\ORM\Mapping as ORM;
+use OrangeHRM\Claim\entity\Decorator\ClaimRequestDecorator;
```
Comment (RajithaKumara) [inline src/plugins/orangehrmClaimPlugin/entity/ClaimRequest.php:24]:
```suggestion
use OrangeHRM\Entity\Decorator\ClaimRequestDecorator;
```

File: src/plugins/orangehrmClaimPlugin/entity/ClaimRequest.php [hunk 18]
```diff
@@ -0,0 +1,301 @@
...
+     * @ORM\JoinColumn(name="event_type_id", referencedColumnName="id")
+     */
+    private ClaimEvent $claimEvent;
+
+    /**
+     * @var string|null
+     *
+     * @ORM\Column(name="description", type="string", length=255)
+     */
+    private ?string $description;
```
Comment (RajithaKumara) [inline src/plugins/orangehrmClaimPlugin/entity/ClaimRequest.php:89]:
```suggestion
    private ?string $description = null;
```

File: src/plugins/orangehrmClaimPlugin/entity/ClaimRequest.php [hunk 19]
```diff
@@ -0,0 +1,301 @@
...
+
+    /**
+     * @var string
+     *
+     * @ORM\Column(name="currency", type="string", length=3)
```
Comment (RajithaKumara) [inline src/plugins/orangehrmClaimPlugin/entity/ClaimRequest.php:94]: isn't this a relational column

File: src/plugins/orangehrmClaimPlugin/entity/Expense.php [hunk 20]
```diff
@@ -0,0 +1,195 @@
...
+namespace OrangeHRM\Entity;
+
+use DateTime;
+use Doctrine\ORM\Mapping as ORM;
+
+/**
+ * @ORM\Table(name="ohrm_expense")
+ * @ORM\Entity
+ */
+class Expense
```
Comment (RajithaKumara) [inline src/plugins/orangehrmClaimPlugin/entity/Expense.php:29]: is this relevant to this PR?

File: src/plugins/orangehrmClaimPlugin/test/Api/ClaimExpenseTypesAPITest.php [hunk 21]
```diff
@@ -0,0 +1,46 @@
...
+ */
+
+namespace OrangeHRM\Tests\Claim\Api;
+
+use OrangeHRM\Claim\Api\ClaimExpenseTypesAPI;
+use OrangeHRM\Framework\Services;
+use OrangeHRM\Tests\Util\EndpointIntegrationTestCase;
+use OrangeHRM\Tests\Util\Integration\TestCaseParams;
+
+class ClaimExpenseTypesAPITest extends EndpointIntegrationTestCase
```
Comment (RajithaKumara) [inline src/plugins/orangehrmClaimPlugin/test/Api/ClaimExpenseTypesAPITest.php:27]:
```suggestion
class ClaimExpenseTypeAPITest extends EndpointIntegrationTestCase
```
add groups

File: src/plugins/orangehrmClaimPlugin/test/Api/ClaimRequestAPITest.php [hunk 22]
```diff
@@ -0,0 +1,46 @@
...
+ */
+
+namespace OrangeHRM\Tests\Claim\Api;
+
+use OrangeHRM\Claim\Api\ClaimRequestAPI;
+use OrangeHRM\Framework\Services;
+use OrangeHRM\Tests\Util\EndpointIntegrationTestCase;
+use OrangeHRM\Tests\Util\Integration\TestCaseParams;
+
+class ClaimRequestAPITest extends EndpointIntegrationTestCase
```
Comment (RajithaKumara) [inline src/plugins/orangehrmClaimPlugin/test/Api/ClaimRequestAPITest.php:27]: add groups

**Review by RajithaKumara — CHANGES_REQUESTED** [review]

File: installer/Migration/V5_4_0/Migration.php (hunk 1 above)
Comment (ManulMax) [reply installer/Migration/V5_4_0/Migration.php:87]: add to migration

File: src/plugins/orangehrmClaimPlugin/Api/ClaimExpenseTypesAPI.php (hunk 2 above)
Comment (ManulMax) [reply src/plugins/orangehrmClaimPlugin/Api/ClaimExpenseTypesAPI.php:70]: added

**Pull request merged**